crates/qasm3/src/expr.rs

```diff
@@ -0,0 +1,387 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::types::PyTuple;
+
+use hashbrown::HashMap;
+
+use oq3_semantics::asg;
+use oq3_semantics::symbols::{SymbolId, SymbolTable};
+use oq3_semantics::types::Type;
+
+use crate::build::PySymbolTable;
+use crate::circuit::PyRegister;
+use crate::error::QASM3ImporterError;
+
+pub fn eval_gate_param(
+    _py: Python,
+    _our_symbols: &PySymbolTable,
+    _ast_symbols: &SymbolTable,
+    param: &asg::TExpr,
+) -> PyResult<f64> {
+    // Only handling float parameters in this first pass of the importer.
+    match param.get_type() {
+        Type::Float(_, is_const) => {
+            if is_const.clone().into() {
+                match param.expression() {
+                    asg::Expr::Literal(asg::Literal::Float(lit)) => {
+                        lit.value().parse().map_err(|_| {
+                            QASM3ImporterError::new_err(format!(
+                                "invalid float literal: '{}'",
+                                lit.value()
+                            ))
+                        })
+                    }
+                    expr => Err(QASM3ImporterError::new_err(format!(
+                        "unhandled expression for floating-point constant: {:?}",
+                        expr
+                    ))),
+                }
+            } else {
+                Err(QASM3ImporterError::new_err(format!(
+                    "expected a constant float, but found a runtime value: {:?}",
+                    param
+                )))
+            }
+        }
+        ty => Err(QASM3ImporterError::new_err(format!(
+            "expected an angle-like type, but saw {:?}",
```

Comment:
This might just be my unfamiliarity with oq3, but won't this match an angle type parameter because it's the catch all? With a function name `eval_gate_param` I assumed an angle type input could be passed to it. So you'd get an error message, "expected an angle-like type, but saw Angle"

Reply:
I was more concerned with the error message here in case an angle type were encountered, I figured there was a reason for the limitations here. But if that can't happen in practice then there's nothing to worry about here.

Reply:
Oh good point, I'll put in a specific catch for if the expression type is `angle`.

Reply:
Done in fcc79dfea2.